src/main/java/io/cryostat/recordings/RecordingTargetHelper.java

```diff
@@ -207,4 +333,48 @@ public IRecordingDescriptor startRecording(
 
         return builder.build();
     }
+
+    private boolean snapshotIsReadable(
+            ConnectionDescriptor connectionDescriptor, InputStream snapshot) throws IOException {
+        if (!targetConnectionManager.markConnectionInUse(connectionDescriptor)) {
+            throw new IOException(
+                    "Target connection unexpectedly closed while streaming recording");
+        }
+
+        try {
+            PushbackInputStream pushbackSnapshot = new PushbackInputStream(snapshot);
+            int b = pushbackSnapshot.read();
+            pushbackSnapshot.unread(b);
```

Comment:
Now that this has been refactored and extracted in this way, I wonder if the `PushbackInputStream` is actually needed at all. This `snapshotIsReadable` method is only called in a branch of `verifySnapshot` that occurs after `verifySnapshot` has also called through to `this.getRecording` on its own - which means it's potentially opening a new target connection (probably just reusing a cached one), but it is definitely opening a new underlying stream of the snapshot recording data in the target JVM. The pushback/unread functionality would only seem to be needed if we were trying to verify that a provided InputStream was non-empty without opening a whole new stream to the same underlying remote recording. If we're opening a new stream to perform the verification then I think we can just use the InputStream as-is and try to read from it (https://docs.oracle.com/javase/7/docs/api/java/io/InputStream.html#read()) - if the return value is `-1` then it was empty, or if an exception was thrown then the connection was closed or the network failed etc.

Please check and test if replacing the PushbackInputStream with simply reading the first byte out of the newly-created stream works, and if so, let's go with that simpler version for this PR and leave the rest of this refactored structure the same. At this point it's just an internal detail of the `snapshotIsReadable` method - everything else seems to be put together well.

Where the PushbackInputStream *would* be extremely useful is in ex. `TargetRecordingGetHandler`. Then the handler could return `404` if there is no such requested recording, or ex. `204` if the recording (including a snapshot recording) could be located in the target but the stream contained 0 bytes. This would definitely be an enhancement for a future PR.

Reply:
That all makes sense, I agree. I'll go ahead and try out `InputStream` instead; I'm fairly certain it'll work out. 

The only thing I'd like to add is that when I originally used `InputStream` instead of `PushbackInputStream`, I was checking to see if `read()` returns `-1` (since logically, an empty recording means the end of the stream will be reached immediately) but what actually happens is that an `IOException` is thrown. Therefore, the code will be the same except for the replacement of the `PushbackInputStream`.

Reply:
Okay, makes sense. That might just be something specific to the `InputStream` implementation we're using under the covers (I think it's this: https://github.com/openjdk/jmc/blob/master/application/org.openjdk.jmc.rjmx.services.jfr/src/main/java/org/openjdk/jmc/rjmx/services/jfr/internal/JfrRecordingInputStreamV2.java). It seems like it "should" be returning `-1`, but there are cases where it throws `IOException: stream closed`, so it seems we fall through that for whatever reason.